pysb/simulator/cupsoda.py

```diff
@@ -0,0 +1,595 @@
+       simulations of mass-action kinetics models with cupSODA, Journal of
+       Supercomputing, 69(1), pp.17-24.
+    2. Petzold, L., 1983. Automatic selection of methods for solving stiff and
+       nonstiff systems of ordinary differential equations. SIAM journal on
+       scientific and statistical computing, 4(1), pp.136-148.
+    """
+
+    _supports = {'multi_initials': True, 'multi_param_values': True}
+
+    _memory_options = {'global': '0', 'shared': '1', 'sharedconstant': '2'}
+
+    default_integrator_options = {
+        # some sane default options for a few well-known integrators
+        'cupsoda': {
+            'max_steps': 20000, # max # of internal iterations (LSODA's MXSTEP)
+            'atol': 1e-8,  # absolute tolerance
+            'rtol': 1e-8,  # relative tolerance
+            'n_blocks': None,  # number of GPU blocks
+            'memory_usage': 'sharedconstant'}}  # see _memory_options dict
+
+    def __init__(self, model, tspan=None, initials=None, param_values=None,
+                 verbose=False, **kwargs):
+        super(CupSodaSimulator, self).__init__(model, tspan=tspan,
+                                               initials=initials,
+                                               param_values=param_values,
+                                               verbose=verbose, **kwargs)
+        self.gpu = kwargs.get('gpu', 0)
+        self._obs_species_only = kwargs.get('obs_species_only', True)
+        self._cleanup = kwargs.get('cleanup', True)
+        self._prefix = kwargs.get('prefix', self._model.name.replace('.', '_'))
+        self._base_dir = kwargs.get('base_dir', None)
+        self.integrator = kwargs.get('integrator', 'cupsoda')
+
+        # generate the equations for the model
+        pysb.bng.generate_equations(self._model, self._cleanup, self.verbose)
+
+        # build integrator options list from our defaults and any kwargs
+        # passed to this function
+        options = {}
+        if self.default_integrator_options.get(self.integrator):
+            options.update(self.default_integrator_options[
+                self.integrator])  # default options
+        else:
+            raise SimulatorException(
+                "Integrator type '" + self.integrator + "' not recognized.")
+        options.update(kwargs.get('integrator_options', {}))  # overwrite
+
+        # defaults
+        self.opts = options
+        self._out_species = None
+
+        # private variables (to reduce the number of function calls)
+        self._len_rxns = len(self._model.reactions)
+        self._len_species = len(self._model.species)
+        self._len_params = len(self._model.parameters)
+        self._model_parameters_rules = self._model.parameters_rules()
+
+        # Set cupsoda verbosity level
+        logger_level = self._logger.logger.getEffectiveLevel()
+        if logger_level > logging.DEBUG:
+            self._cupsoda_verbose = 0
+        elif logger_level <= 5:
```

Comment:
I think it might be more readable to reorganize this so the tested ranges progress in order rather than skipping around. Also to document what level 5 is via a named constant (no magic numbers). Something like this:

MAX_DEBUG = 5 (at the top level)
...
if level <= MAX_DEBUG ... elif level <= logging.DEBUG ... else ...